From patchwork Tue Jan 20 18:09:04 2026
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Alexey Charkov <alchark@gmail.com>
X-Patchwork-Id: 2186923
X-Patchwork-Delegate: narmstrong@baylibre.com
Return-Path: <u-boot-bounces@lists.denx.de>
X-Original-To: incoming@patchwork.ozlabs.org
Delivered-To: patchwork-incoming@legolas.ozlabs.org
Authentication-Results: legolas.ozlabs.org;
	dkim=pass (2048-bit key;
 unprotected) header.d=gmail.com header.i=@gmail.com header.a=rsa-sha256
 header.s=20230601 header.b=T1RIXbeb;
	dkim-atps=neutral
Authentication-Results: legolas.ozlabs.org;
 spf=pass (sender SPF authorized) smtp.mailfrom=lists.denx.de
 (client-ip=85.214.62.61; helo=phobos.denx.de;
 envelope-from=u-boot-bounces@lists.denx.de; receiver=patchwork.ozlabs.org)
Received: from phobos.denx.de (phobos.denx.de [85.214.62.61])
	(using TLSv1.3 with cipher TLS_AES_256_GCM_SHA384 (256/256 bits)
	 key-exchange x25519 server-signature ECDSA (secp384r1) server-digest SHA384)
	(No client certificate requested)
	by legolas.ozlabs.org (Postfix) with ESMTPS id 4dwdhB1x2Kz1xsN
	for <incoming@patchwork.ozlabs.org>; Wed, 21 Jan 2026 07:09:14 +1100 (AEDT)
Received: from h2850616.stratoserver.net (localhost [IPv6:::1])
	by phobos.denx.de (Postfix) with ESMTP id 4768382A9E;
	Tue, 20 Jan 2026 21:09:09 +0100 (CET)
Authentication-Results: phobos.denx.de;
 dmarc=pass (p=none dis=none) header.from=gmail.com
Authentication-Results: phobos.denx.de;
 spf=pass smtp.mailfrom=u-boot-bounces@lists.denx.de
Authentication-Results: phobos.denx.de;
	dkim=pass (2048-bit key;
 unprotected) header.d=gmail.com header.i=@gmail.com header.b="T1RIXbeb";
	dkim-atps=neutral
Received: by phobos.denx.de (Postfix, from userid 109)
 id AD25083015; Tue, 20 Jan 2026 20:48:30 +0100 (CET)
X-Spam-Checker-Version: SpamAssassin 3.4.2 (2018-09-13) on phobos.denx.de
X-Spam-Level: 
X-Spam-Status: No, score=-1.1 required=5.0 tests=BAYES_00,DKIM_SIGNED,
 DKIM_VALID,DKIM_VALID_AU,DKIM_VALID_EF,FORGED_GMAIL_RCVD,FREEMAIL_FROM,
 RCVD_IN_DNSWL_BLOCKED,SPF_HELO_NONE,SPF_PASS autolearn=no
 autolearn_force=no version=3.4.2
Received: from mail-ej1-x62b.google.com (mail-ej1-x62b.google.com
 [IPv6:2a00:1450:4864:20::62b])
 (using TLSv1.3 with cipher TLS_AES_128_GCM_SHA256 (128/128 bits))
 (No client certificate requested)
 by phobos.denx.de (Postfix) with ESMTPS id 595A08063E
 for <u-boot@lists.denx.de>; Tue, 20 Jan 2026 20:48:28 +0100 (CET)
Authentication-Results: phobos.denx.de;
 dmarc=pass (p=none dis=none) header.from=gmail.com
Authentication-Results: phobos.denx.de;
 spf=pass smtp.mailfrom=alchark@gmail.com
Received: by mail-ej1-x62b.google.com with SMTP id
 a640c23a62f3a-b8715a4d9fdso803596566b.0
 for <u-boot@lists.denx.de>; Tue, 20 Jan 2026 11:48:28 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=gmail.com; s=20230601; t=1768938508; x=1769543308; darn=lists.denx.de;
 h=cc:to:in-reply-to:references:message-id:content-transfer-encoding
 :mime-version:subject:date:from:from:to:cc:subject:date:message-id
 :reply-to; bh=UCtp+N2YjrYVHtkTDwdRwi4lBZlAWlUqmgmD9IoPzik=;
 b=T1RIXbebr5ET2CdwTnkMkLyKHgbTZABHd0ocSIZopze78EXBfT5qp6WfRXe7Lgl/IZ
 Yu2fu2c0f/IXyLkdNQQiDElH+lOGvREqHz0g5xTa4TKJ52lGooJp4Y3R+nTCiaRCRpgB
 j5Ea3uAmU8/vVWYhIM/58Txo6oaP2D6G+v/Gvwuecz7isjfaJTrpBWFy98szkVlXl0/4
 25CEckNkb0WBqS0FSTV+RzleGig/kIkN5OlzBNpa21Yz/OrRXn3oi1NZls9Q9BK3sZmC
 BFHHcsuv9qT0XEABs616IFtY4TOqFXZw9+7NzkQ+zREa5TkpQ4szfVFnKyvV3cYuvCAO
 Ge/A==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=1e100.net; s=20230601; t=1768938508; x=1769543308;
 h=cc:to:in-reply-to:references:message-id:content-transfer-encoding
 :mime-version:subject:date:from:x-gm-gg:x-gm-message-state:from:to
 :cc:subject:date:message-id:reply-to;
 bh=UCtp+N2YjrYVHtkTDwdRwi4lBZlAWlUqmgmD9IoPzik=;
 b=Ab7Cz4z4wzDB1ziEL4izj6OkYhHVKTpiECZjCdD1tHPHLbcuzu99HmKBru9NhGLaLf
 D9i3ZwsubkunFpI7//P497a8bZ8k4DyozxAOkNo86mqr8x+e8wEH/z9WDsEHmc6cdYgw
 QeOTrVK4unjCLS5RIzolyrqrtUEzvMb/4kAea4UzlLXxzaAgTpmRXv/XANaHTtsoFDaL
 1gRM1Q7D9rrTxEki1YRyZ6z2d3xfDgjBbBuXTDgw3jCPdo3GZmUqwilU0NMs3F/B6Wme
 T+Ez9/9YheQ4hmkYN+qvt+R9D1YQMV3AVq/R6MJEAQeEHo4O37z7vltcxEkSfP7JDmde
 1HYw==
X-Gm-Message-State: AOJu0YybpuDwF8nWTMZio1mjxSWO6JxjaI7iVkDeToUaCN3XAE1TrIgv
 Fzzrfh0jIBxuM/sZktMEnbLKoBkGNQ8sN+5Z9cteOCJXZSE7fB/klspsczHcjI6HPkI=
X-Gm-Gg: AZuq6aL/Hm5uzaMwrWfrUwPWGB+H2ScZQ/WpeKvbRbYVJjRqGYAmN63atnDBau2M50b
 RDw2Ux3o14eE2xU+XAgljflbcdQxGExUVLaLSKvsA04Vnk7TaqqPNbXNNcIkBGZPHj09Qa2wq2X
 qFPZXJGpW0BxuGkCa54B3Yj/ePoY0AiJn7xS0b7kPmZnysPC1NVvqZzPZTejOXdyVwMXasw0hPw
 5GajKEbzQ0mw1nKCKrVV3bNc0MehtkKbXG/XUfaIvBV0Uho00582+ZAMndum7Y/dABD0jW1yaGo
 WdIXvtsJ0TPEqACFdJnnnIPqtrw40R0OoUsNXL+8hCklsMBVKtbF+HkWCiVhIg8Ve/E3u/XQhnM
 hm/gIo84zzAJMxtpVx8WKUutfjPnnrjmTj0j5SbvH4kMDbTgWWRwaTJSONIw7TQ4f+oewqr41P2
 XAsMlER32n1aryPN5B1kM2AVdFn6h25eMyxzv7iR3/Hbx2xP5UIG/xbvW3v4icK84=
X-Received: by 2002:a05:6000:2511:b0:42f:dbbc:5103 with SMTP id
 ffacd0b85a97d-4356a07728fmr21223138f8f.35.1768932584783;
 Tue, 20 Jan 2026 10:09:44 -0800 (PST)
Received: from alchark-surface.localdomain (bba-83-110-134-52.alshamil.net.ae.
 [83.110.134.52]) by smtp.gmail.com with ESMTPSA id
 ffacd0b85a97d-43569998240sm30168043f8f.43.2026.01.20.10.09.37
 (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
 Tue, 20 Jan 2026 10:09:44 -0800 (PST)
From: Alexey Charkov <alchark@gmail.com>
Date: Tue, 20 Jan 2026 22:09:04 +0400
Subject: [PATCH v5 3/4] ufs: rockchip: Add device reset support
MIME-Version: 1.0
Message-Id: <20260120-rk3576-ufs-v5-3-0edb61b301b7@gmail.com>
References: <20260120-rk3576-ufs-v5-0-0edb61b301b7@gmail.com>
In-Reply-To: <20260120-rk3576-ufs-v5-0-0edb61b301b7@gmail.com>
To: u-boot@lists.denx.de, Shawn Lin <shawn.lin@rock-chips.com>
Cc: Tom Rini <trini@konsulko.com>, Kever Yang <kever.yang@rock-chips.com>,
 Jonas Karlman <jonas@kwiboo.se>, Heiko Stuebner <heiko@sntech.de>,
 Neil Armstrong <neil.armstrong@linaro.org>,
 Bhupesh Sharma <bhupesh.linux@gmail.com>,
 Neha Malcom Francis <n-francis@ti.com>, Greg Malysa <malysagreg@gmail.com>,
 Vasileios Bimpikas <vasileios.bimpikas@analog.com>,
 Arturs Artamonovs <arturs.artamonovs@analog.com>,
 Utsav Agarwal <utsav.agarwal@analog.com>,
 Ian Roberts <ian.roberts@timesys.com>,
 Nathan Barrett-Morrison <nathan.morrison@timesys.com>,
 Peng Fan <peng.fan@nxp.com>,
 "Kory Maincent (TI.com)" <kory.maincent@bootlin.com>,
 Alif Zakuan Yuslaimi <alif.zakuan.yuslaimi@altera.com>,
 Simon Glass <sjg@chromium.org>, Yao Zi <me@ziyao.cc>,
 Stefan Roese <stefan.roese@mailbox.org>,
 Mattijs Korpershoek <mkorpershoek@kernel.org>,
 Sumit Garg <sumit.garg@kernel.org>, Heiko Schocher <hs@nabladev.com>,
 Anshul Dalal <anshuld@ti.com>, Andre Przywara <andre.przywara@arm.com>,
 Heinrich Schuchardt <xypron.glpk@gmx.de>,
 Quentin Schulz <quentin.schulz@cherry.de>, Andrew Davis <afd@ti.com>,
 Hrushikesh Salunke <h-salunke@ti.com>,
 Dario Binacchi <dario.binacchi@amarulasolutions.com>,
 Dinesh Maniyam <dinesh.maniyam@altera.com>,
 Ilias Apalodimas <ilias.apalodimas@linaro.org>,
 Philippe Reynes <philippe.reynes@softathome.com>,
 Marek Vasut <marek.vasut@mailbox.org>,
 Philipp Tomsich <philipp.tomsich@vrull.eu>,
 Marek Vasut <marek.vasut+renesas@mailbox.org>,
 Igor Belwon <igor.belwon@mentallysanemainliners.org>,
 Tuyen Dang <tuyen.dang.xa@renesas.com>,
 Xuhui Lin <xuhui.lin@rock-chips.com>, Johan Jonker <jbx6244@gmail.com>,
 Lukasz Czechowski <lukasz.czechowski@thaumatec.com>,
 Dhruva Gole <d-gole@ti.com>, Patrice Chotard <patrice.chotard@foss.st.com>,
 Paul Barker <paul.barker.ct@bp.renesas.com>,
 Gabriel Fernandez <gabriel.fernandez@foss.st.com>,
 Leo Yu-Chi Liang <ycliang@andestech.com>,
 Christian Marangi <ansuelsmth@gmail.com>,
 Elaine Zhang <zhangqing@rock-chips.com>,
 Joseph Chen <chenjh@rock-chips.com>, Alexey Charkov <alchark@gmail.com>
X-Mailer: b4 0.14.3
X-Developer-Signature: v=1; a=openpgp-sha256; l=3401; i=alchark@gmail.com;
 h=from:subject:message-id; bh=ZOzJH7cBYN8OgyVAK1Gav+v1EmvlG6lr57/Bo98n9/E=;
 b=owGbwMvMwCW2adGNfoHIK0sZT6slMWTmHzl1O4f/qe3ljYsTxUJzJnEW7vo5rZhpZSzbPCmrX
 92vk/PMOiayMIhxMViKKbLM/bbEdqoR36xdHh5fYeawMoEMkRZpYAACFga+3MS8UiMdIz1TbUM9
 Q0MdYx0jBi5OAZjqGVaMDA9ML70WqWmINcs24513YMPW/T9zmvYeqBe5Itwiap0W/5Hhv1ua96L
 fy1vsyhK5T6Wb2O6KiN7ewGWu6CSjcKej8CIzHwA=
X-Developer-Key: i=alchark@gmail.com; a=openpgp;
 fpr=9DF6A43D95320E9ABA4848F5B2A2D88F1059D4A5
X-Mailman-Approved-At: Tue, 20 Jan 2026 21:09:08 +0100
X-BeenThere: u-boot@lists.denx.de
X-Mailman-Version: 2.1.39
Precedence: list
List-Id: U-Boot discussion <u-boot.lists.denx.de>
List-Unsubscribe: <https://lists.denx.de/options/u-boot>,
 <mailto:u-boot-request@lists.denx.de?subject=unsubscribe>
List-Archive: <https://lists.denx.de/pipermail/u-boot/>
List-Post: <mailto:u-boot@lists.denx.de>
List-Help: <mailto:u-boot-request@lists.denx.de?subject=help>
List-Subscribe: <https://lists.denx.de/listinfo/u-boot>,
 <mailto:u-boot-request@lists.denx.de?subject=subscribe>
Errors-To: u-boot-bounces@lists.denx.de
Sender: "U-Boot" <u-boot-bounces@lists.denx.de>
X-Virus-Scanned: clamav-milter 0.103.8 at phobos.denx.de
X-Virus-Status: Clean

Wire up the GPIO line which Rockchip RK3576 UFS controller uses to reset
the connected UFS device.

This seems necessary at least for some UFS modules and fixes the following
error while enumerating UFS storage:

ufshcd-rockchip ufshc@2a2d0000: ufshcd_link_startup: Device not present
ufshcd-rockchip ufshc@2a2d0000: link startup failed -6
ufshcd-rockchip ufshc@2a2d0000: ufshcd_pltfrm_init() failed -6

Note that the GPIO descriptor for device resets is already required by the
DT binding (link enclosed).

Link: https://elixir.bootlin.com/linux/v6.18.5/source/Documentation/devicetree/bindings/ufs/rockchip,rk3576-ufshc.yaml#L70
Fixes: 76465ce21ee4 ("ufs: rockchip: Add initial support")
Reviewed-by: Neil Armstrong <neil.armstrong@linaro.org>
Reviewed-by: Shawn Lin <shawn.lin@rock-chips.com>
Signed-off-by: Alexey Charkov <alchark@gmail.com>
---
 drivers/ufs/Kconfig        |  4 ++++
 drivers/ufs/ufs-rockchip.c | 22 ++++++++++++++++++++++
 drivers/ufs/ufs-rockchip.h |  1 +
 3 files changed, 27 insertions(+)

diff --git a/drivers/ufs/Kconfig b/drivers/ufs/Kconfig
index 6c75bb2a0790..49472933de36 100644
--- a/drivers/ufs/Kconfig
+++ b/drivers/ufs/Kconfig
@@ -76,6 +76,10 @@ config UFS_RENESAS_GEN5
 config UFS_ROCKCHIP
 	bool "Rockchip specific hooks to UFS controller platform driver"
 	depends on UFS
+	depends on DM_GPIO
+	depends on RESET_ROCKCHIP
+	depends on SPL_DM_GPIO || !SPL_UFS_SUPPORT
+	depends on SPL_RESET_ROCKCHIP || !SPL_UFS_SUPPORT
 	help
 	  This selects the Rockchip specific additions to UFSHCD platform driver.
 
diff --git a/drivers/ufs/ufs-rockchip.c b/drivers/ufs/ufs-rockchip.c
index 0384244387da..3576f6e7a2b6 100644
--- a/drivers/ufs/ufs-rockchip.c
+++ b/drivers/ufs/ufs-rockchip.c
@@ -5,6 +5,7 @@
  * Copyright (C) 2025 Rockchip Electronics Co.Ltd.
  */
 
+#include <asm/gpio.h>
 #include <asm/io.h>
 #include <clk.h>
 #include <dm.h>
@@ -153,11 +154,31 @@ static int ufs_rockchip_common_init(struct ufs_hba *hba)
 		return err;
 	}
 
+	err = gpio_request_by_name(dev, "reset-gpios", 0, &host->device_reset,
+				   GPIOD_IS_OUT | GPIOD_ACTIVE_LOW);
+	if (err) {
+		dev_err(dev, "Cannot get reset GPIO\n");
+		return err;
+	}
+
 	host->hba = hba;
 
 	return 0;
 }
 
+static int ufs_rockchip_device_reset(struct ufs_hba *hba)
+{
+	struct ufs_rockchip_host *host = dev_get_priv(hba->dev);
+
+	dm_gpio_set_value(&host->device_reset, true);
+	udelay(20);
+
+	dm_gpio_set_value(&host->device_reset, false);
+	udelay(20);
+
+	return 0;
+}
+
 static int ufs_rockchip_rk3576_init(struct ufs_hba *hba)
 {
 	int ret = 0;
@@ -175,6 +196,7 @@ static struct ufs_hba_ops ufs_hba_rk3576_vops = {
 	.init = ufs_rockchip_rk3576_init,
 	.phy_initialization = ufs_rockchip_rk3576_phy_init,
 	.hce_enable_notify = ufs_rockchip_hce_enable_notify,
+	.device_reset = ufs_rockchip_device_reset,
 };
 
 static const struct udevice_id ufs_rockchip_of_match[] = {
diff --git a/drivers/ufs/ufs-rockchip.h b/drivers/ufs/ufs-rockchip.h
index 3dcb80f57020..50c2539da78a 100644
--- a/drivers/ufs/ufs-rockchip.h
+++ b/drivers/ufs/ufs-rockchip.h
@@ -72,6 +72,7 @@ struct ufs_rockchip_host {
 	void __iomem *ufs_sys_ctrl;
 	void __iomem *mphy_base;
 	struct reset_ctl_bulk rsts;
+	struct gpio_desc device_reset;
 	struct clk ref_out_clk;
 	uint64_t caps;
 	uint32_t phy_config_mode;
